### PR TITLE
[Perf] Optimize cutlass moe problem size calculation, 5.3% E2E Throughput improvement, 2.2% TTFT improvement

### PR DESCRIPTION
**Pull request by yewentao256**
## Purpose

Part of the https://github.com/vllm-project/vllm/issues/31755

Here we add a kernel for faster calculation of problem size

## Test

`export MODEL="zai-org/GLM-4.7-FP8"`

`vllm serve $MODEL -tp 8   --port 9256 --enable-expert-parallel --max_num_seqs 128`

### Acc

`lm_eval --model local-completions --model_args "base_url=http://127.0.0.1:9256/v1/completions,model=$MODEL,num_concurrent=1024" --tasks gsm8k`

```bash
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.9469|±  |0.0062|
|     |       |strict-match    |     5|exact_match|↑  |0.9462|±  |0.0062|
```

With EPLB:

`vllm serve $MODEL -tp 8   --port 9256 --enable-expert-parallel --max_num_seqs 128 --enable-eplb`

```bash
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.9462|±  |0.0062|
|     |       |strict-match    |     5|exact_match|↑  |0.9454|±  |0.0063|
```

### Perf

`vllm bench serve --model $MODEL  --dataset-name random --host 127.0.0.1 --port 9256 --random-input-len 2 --random-output-len 128 --request-rate inf --num-prompts 128`

```bash
# Now
============ Serving Benchmark Result ============
Successful requests:                     128       
Failed requests:                         0         
Benchmark duration (s):                  4.79      
Total input tokens:                      256       
Total generated tokens:                  16384     
Request throughput (req/s):              26.71     
Output token throughput (tok/s):         3419.28   
Peak output token throughput (tok/s):    3584.00   
Peak concurrent requests:                128.00    
Total token throughput (tok/s):          3472.71   
---------------Time to First Token----------------
Mean TTFT (ms):                          165.77    
Median TTFT (ms):                        158.45    
P99 TTFT (ms):                           187.77    
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          36.17     
Median TPOT (ms):                        36.17     
P99 TPOT (ms):                           36.22     
---------------Inter-token Latency----------------
Mean ITL (ms):                           36.17     
Median ITL (ms):                         36.17     
P99 ITL (ms):                            39.66     
==================================================
# main 
============ Serving Benchmark Result ============
Successful requests:                     128       
Failed requests:                         0         
Benchmark duration (s):                  5.05      
Total input tokens:                      256       
Total generated tokens:                  16384     
Request throughput (req/s):              25.37     
Output token throughput (tok/s):         3246.74   
Peak output token throughput (tok/s):    3456.00   
Peak concurrent requests:                128.00    
Total token throughput (tok/s):          3297.47   
---------------Time to First Token----------------
Mean TTFT (ms):                          169.53    
Median TTFT (ms):                        160.36    
P99 TTFT (ms):                           196.47    
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          38.14     
Median TPOT (ms):                        38.17     
P99 TPOT (ms):                           38.20     
---------------Inter-token Latency----------------
Mean ITL (ms):                           38.14     
Median ITL (ms):                         38.08     
P99 ITL (ms):                            42.53     
==================================================
```

---

> [!NOTE]
> Introduces a faster path for CUTLASS MoE problem-size setup and wires it through C++/Python to fused MoE.
> 
> - New CUDA kernel + entry points: `compute_problem_sizes_from_expert_offsets` with callers and registry (`moe_data.cu`, `scaled_mm_entry.cu`, `torch_bindings.cpp`) and Python API `ops.get_cutlass_moe_mm_problem_sizes_from_expert_offsets`
> - Replace `topk_ids` counting with expert-offset–based sizing in `cutlass_moe.py` (both FP8 and W4A8 paths): allocate `problem_sizes{1,2}` for `local_E`, call the new op, derive `expert_offsets` from `expert_first_token_offset`, and always pass `expert_first_token_offset` to `moe_unpermute`
> - Swap-AB control via size heuristic or forced for RS GEMM; use `VLLM_DISPATCH_BOOL` for launch-time specialization; minor pointer/type cleanups
> - Public headers updated (`ops.h`) and Python wrappers added in `_custom_ops.py`
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 2c089dabfcd6d6de623ecae6560821e12859d907. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

---

> [!NOTE]
> <sup>[Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) is generating a summary for commit 19024818a346f34faae2a97d1294ae740532ec61. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: csrc/quantization/w8a8/cutlass/moe/moe_data.cu [hunk 1]
```diff
@@ -153,6 +153,99 @@ void get_cutlass_moe_mm_problem_sizes_caller(
...
+    int32_t* __restrict__ problem_sizes1, int32_t* __restrict__ problem_sizes2,
+    const int num_experts, const int n, const int k) {
+  int const expert_id = blockIdx.x * blockDim.x + threadIdx.x;
+  if (expert_id >= num_experts) {
+    return;
+  }
+
+  int64_t const m64 = expert_first_token_offset[expert_id + 1] -
+                      expert_first_token_offset[expert_id];
+  int32_t const m = static_cast<int32_t>(m64);
```
[the remark this answers is not in the document]
Comment (yewentao256) [reply csrc/quantization/w8a8/cutlass/moe/moe_data.cu:165]: For moe, nearly never be true of the overflow

File: csrc/quantization/w8a8/cutlass/moe/moe_data.cu [hunk 2]
```diff
@@ -153,6 +153,99 @@ void get_cutlass_moe_mm_problem_sizes_caller(
...
+
+  int const threads = 256;
+  int const blocks = (num_experts + threads - 1) / threads;
+
+  auto const* offsets_ptr =
+      static_cast<const int64_t*>(expert_first_token_offset.data_ptr());
+  auto* ps1_ptr = static_cast<int32_t*>(problem_sizes1.data_ptr());
+  auto* ps2_ptr = static_cast<int32_t*>(problem_sizes2.data_ptr());
+
+  if (swap_ab) {
```
Comment (ProExpertProg) [inline csrc/quantization/w8a8/cutlass/moe/moe_data.cu]: Use dispatch boolean macro

File: csrc/quantization/w8a8/cutlass/moe/moe_data.cu [hunk 3]
```diff
@@ -153,6 +153,99 @@ void get_cutlass_moe_mm_problem_sizes_caller(
...
+  auto stream = at::cuda::getCurrentCUDAStream(
+      expert_first_token_offset.device().index());
+
+  int const threads = 256;
+  int const blocks = (num_experts + threads - 1) / threads;
+
+  auto const* offsets_ptr =
+      static_cast<const int64_t*>(expert_first_token_offset.data_ptr());
+  auto* ps1_ptr = static_cast<int32_t*>(problem_sizes1.data_ptr());
+  auto* ps2_ptr = static_cast<int32_t*>(problem_sizes2.data_ptr());
```
Comment (ProExpertProg) [inline csrc/quantization/w8a8/cutlass/moe/moe_data.cu]: Does this work?
```suggestion
  auto const* offsets_ptr = expert_first_token_offset.data_ptr<int64_t>();
  auto* ps1_ptr = problem_sizes1.data_ptr<int32_t>();
  auto* ps2_ptr = problem_sizes2.data_ptr<int32_t>();
```

**Review by ProExpertProg — COMMENTED** [review]
This seems like a good opportunity for a Triton kernel, did you try that?

File: vllm/model_executor/layers/fused_moe/cutlass_moe.py [hunk 4]
```diff
@@ -108,15 +108,7 @@ def run_cutlass_moe_fp8(
     assert global_num_experts != -1
     assert a1q_scale is not None
 
-    if expert_map is not None:
-        "Translate info from expert_map to topk_ids"
-        local_topk_ids = torch.where(
-            expert_map[topk_ids] != -1, expert_map[topk_ids], -1
-        )
```
Comment (mgoin) [inline vllm/model_executor/layers/fused_moe/cutlass_moe.py:115]: Please verify correctness for removing the expert_map logic here. I assume this works because moe_permute already handles the mapping, but I'm not sure. I think you should test accuracy with EP and EPLB to properly exercise this case

File: vllm/model_executor/layers/fused_moe/cutlass_moe.py [hunk 5]
```diff
@@ -182,11 +170,11 @@ def run_cutlass_moe_fp8(
             expert_map,
             permuted_hidden_states=a1q_perm,
         )
-        expert_offsets = expert_first_token_offset[:-1]
-
-        ops.get_cutlass_moe_mm_problem_sizes(
-            local_topk_ids, problem_sizes1, problem_sizes2, global_num_experts, N, K
+        swap_ab = a1q.size(0) <= 64
```
Comment (mgoin) [inline vllm/model_executor/layers/fused_moe/cutlass_moe.py:174]: nit: since this is an optimization choice within the kernel, I'd rather the swap_ab decision be made inside the get_cutlass_moe_mm_problem_sizes_from_expert_offsets function. However I see you must force it in the other case. So maybe just a comment on why this heuristic was chosen would be good

File: csrc/quantization/w8a8/cutlass/moe/moe_data.cu [hunk 6]
```diff
@@ -153,6 +150,93 @@ void get_cutlass_moe_mm_problem_sizes_caller(
...
+              "expert_first_token_offset must have num_experts + 1 elements");
+  TORCH_CHECK(num_experts64 <= INT32_MAX, "num_experts must fit in int32");
+  TORCH_CHECK(n <= INT32_MAX && k <= INT32_MAX, "n and k must fit in int32");
+
+  int const num_experts = static_cast<int>(num_experts64);
+  auto stream = at::cuda::getCurrentCUDAStream(
+      expert_first_token_offset.device().index());
+
+  int const threads = 256;
+  int const blocks = (num_experts + threads - 1) / threads;
```
Comment (mgoin) [inline csrc/quantization/w8a8/cutlass/moe/moe_data.cu]: nit: this seems like overkill since each thread only handles one expert, iiuc. could it be num_experts threads in a single block?

File: vllm/model_executor/layers/fused_moe/cutlass_moe.py (hunk 5 above)
Comment (yewentao256) [reply vllm/model_executor/layers/fused_moe/cutlass_moe.py:174]: Fixed

File: csrc/quantization/w8a8/cutlass/moe/moe_data.cu (hunk 6 above)
Comment (yewentao256) [reply csrc/quantization/w8a8/cutlass/moe/moe_data.cu]: Fixed

File: vllm/model_executor/layers/fused_moe/cutlass_moe.py (hunk 4 above)
Comment (yewentao256) [reply vllm/model_executor/layers/fused_moe/cutlass_moe.py:115]:
```bash
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.9462|±  |0.0062|
|     |       |strict-match    |     5|exact_match|↑  |0.9454|±  |0.0063|
```
Tested with EPLB, added in the PR description as well

**Review by mgoin — APPROVED** [review]
Thanks, LGTM!

File: vllm/model_executor/layers/fused_moe/cutlass_moe.py [hunk 7]
```diff
@@ -182,11 +170,12 @@ def run_cutlass_moe_fp8(
...
             permuted_hidden_states=a1q_perm,
         )
-        expert_offsets = expert_first_token_offset[:-1]
-
-        ops.get_cutlass_moe_mm_problem_sizes(
-            local_topk_ids, problem_sizes1, problem_sizes2, global_num_experts, N, K
+        # swap_ab is a CUTLASS grouped-GEMM optimization (M <= 64 reduces padding).
+        swap_ab = a1q.size(0) <= 64
+        ops.get_cutlass_moe_mm_problem_sizes_from_expert_offsets(
+            expert_first_token_offset, problem_sizes1, problem_sizes2, N, K, swap_ab
```
Comment (pavanimajety) [inline vllm/model_executor/layers/fused_moe/cutlass_moe.py:176]: could we use it or cutlass moe fp4 too?

Comment (yewentao256) [reply vllm/model_executor/layers/fused_moe/cutlass_moe.py:176]: Seems out of this PR's scope, I can test it and if could be used, I will have a following up PR for that.
